driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java

```diff
@@ -0,0 +1,277 @@
+        return withLock(lock, () -> {
+            State localState = state;
+            if (!localState.operable()) {
+                return false;
+            } else if (localState == State.IDLE) {
+                state = State.OPERATION_IN_PROGRESS;
+                return true;
+            } else if (localState == State.OPERATION_IN_PROGRESS) {
+                throw new IllegalStateException(MESSAGE_IF_CONCURRENT_OPERATION);
+            } else {
+                throw fail(state.toString());
+            }
+        });
+    }
+
+    /**
+     * Thread-safe.
+     */
+    void endOperation() {
+        boolean doClose = withLock(lock, () -> {
+            State localState = state;
+            if (localState == State.OPERATION_IN_PROGRESS) {
+                state = State.IDLE;
+            } else if (localState == State.CLOSE_PENDING) {
+                state = State.CLOSED;
+                return true;
+            } else if (localState != State.CLOSED) {
+                throw fail(localState.toString());
+            }
+            return false;
+        });
+        if (doClose) {
+            doClose();
+        }
+    }
+
+    /**
+     * Thread-safe.
+     */
+    void close() {
+        boolean doClose = withLock(lock, () -> {
+            State localState = state;
+            if (localState == State.OPERATION_IN_PROGRESS) {
+                state = State.CLOSE_PENDING;
+            } else if (localState != State.CLOSED) {
+                state = State.CLOSED;
+                return true;
+            }
+            return false;
+        });
+        if (doClose) {
+            doClose();
+        }
+    }
+
+    /**
+     * This method is never executed concurrently with either itself or other operations
+     * demarcated by {@link #tryStartOperation()}/{@link #endOperation()}.
+     */
+    abstract void doClose();
+
+    void onCorruptedConnection(@Nullable final C corruptedConnection, final MongoSocketException e) {
+        // if `pinnedConnection` is corrupted, then we cannot kill `serverCursor` via such a connection
+        C localPinnedConnection = pinnedConnection;
+        if (localPinnedConnection != null) {
+            if (corruptedConnection != localPinnedConnection) {
+                e.addSuppressed(new AssertionError("Corrupted connection does not equal the pinned connection."));
+            }
+            skipReleasingServerResourcesOnClose = true;
+        }
+    }
+
+    /**
+     * Thread-safe.
+     */
+    @Nullable
+    ServerCursor getServerCursor() {
+        return serverCursor;
+    }
+
+    void setServerCursor(@Nullable final ServerCursor serverCursor) {
+        assertTrue(state.inProgress());
+        assertNotNull(this.serverCursor);
+        // without `connectionSource` we will not be able to kill `serverCursor` later
+        assertNotNull(connectionSource);
+        this.serverCursor = serverCursor;
+        if (serverCursor == null) {
+            releaseClientResources();
```

Comment:
```suggestion
            close();
```

Reply:
Had to revert as cannot do it for both sync and async cursors.